Doc-tests fail when setting target via CARGO_BUILD_TARGET

I am building against x86_64-unknown-linux-musl by setting CARGO_BUILD_TARGET. The build works. However the doc tests do not pick up the correct target and fail with:

```
error[E0461]: couldn't find crate `error_chain` with expected target triple x86_64-unknown-linux-gnu
```

Notice that the target triple ends with gnu, not musl.

Here is CARGO_BUILD_TARGET:

```
$ env | grep CARGO_BUILD_TARGET
CARGO_BUILD_TARGET=x86_64-unknown-linux-musl
```

Non-doc tests pass, but the doc-tests fail:

```
$ cargo test --release
    Finished release [optimized] target(s) in 0.0 secs
     Running target/x86_64-unknown-linux-musl/release/deps/collins-ba53e34b65340f0f

running 2 tests
test tests::it_encodes_empty_params ... ok
test tests::it_encodes_params ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured

     Running target/x86_64-unknown-linux-musl/release/collins-b486d31bff2ba01a

running 1 test
test it_should_manage_asset ... ignored

test result: ok. 0 passed; 0 failed; 1 ignored; 0 measured

   Doc-tests collins
error[E0461]: couldn't find crate `error_chain` with expected target triple x86_64-unknown-linux-gnu
  --> /home/rust/src/src/lib.rs:11:1
   |
11 | extern crate error_chain;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: crate `error_chain`, path #1, triple x86_64-unknown-linux-musl: /home/rust/src/target/x86_64-unknown-linux-musl/release/deps/liberror_chain-554f08ea2cb4f0f5.rlib

error: aborting due to previous error

thread '<unnamed>' panicked at 'Box<Any>', ../src/librustc_errors/lib.rs:672
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: test failed
```

If I use the --target flag, the doc tests pass.

```
$ cargo test --release --target x86_64-unknown-linux-musl
    Finished release [optimized] target(s) in 0.0 secs
     Running target/x86_64-unknown-linux-musl/release/deps/collins-ba53e34b65340f0f

running 2 tests
test tests::it_encodes_empty_params ... ok
test tests::it_encodes_params ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured

     Running target/x86_64-unknown-linux-musl/release/collins-b486d31bff2ba01a

running 1 test
test it_should_manage_asset ... ignored

test result: ok. 0 passed; 0 failed; 1 ignored; 0 measured
```

Software versions are:

```
$ rustc --version --verbose
rustc 1.12.0 (3191fbae9 2016-09-23)
binary: rustc
commit-hash: 3191fbae9da539442351f883bdabcad0d72efcb6
commit-date: 2016-09-23
host: x86_64-unknown-linux-gnu
release: 1.12.0
$ rustup --version
rustup 0.6.3 (a0e2132 2016-08-28)
$ cargo --version
cargo 0.13.0-nightly (109cb7c 2016-08-19)
```

--- a/src/cargo/ops/cargo_rustc/compilation.rs
+++ b/src/cargo/ops/cargo_rustc/compilation.rs
@@ -44,6 +44,8 @@ pub struct Compilation<'cfg> {
     /// Features enabled during this compilation.
     pub cfgs: HashSet<String>,
 
+    pub target: String,
+
     config: &'cfg Config,
 }
 
@@ -60,6 +62,7 @@ impl<'cfg> Compilation<'cfg> {
             to_doc_test: Vec::new(),
             cfgs: HashSet::new(),
             config: config,
+            target: String::new(),
         }
     }
 

--- a/src/cargo/ops/cargo_rustc/mod.rs
+++ b/src/cargo/ops/cargo_rustc/mod.rs
@@ -159,6 +159,7 @@ pub fn compile_targets<'a, 'cfg: 'a>(ws: &Workspace<'cfg>,
             cx.compilation.native_dirs.insert(dir.clone());
         }
     }
+    cx.compilation.target = cx.target_triple().to_string();
     Ok(cx.compilation)
 }
 

--- a/src/cargo/ops/cargo_test.rs
+++ b/src/cargo/ops/cargo_test.rs
@@ -116,10 +116,8 @@ fn run_doc_tests(options: &TestOptions,
     let config = options.compile_opts.config;
 
     // We don't build/rust doctests if target != host
-    if let Some(target) = options.compile_opts.target {
-        if try!(config.rustc()).host != target {
-            return Ok(errors);
-        }
+    if try!(config.rustc()).host != compilation.target {
+        return Ok(errors);
     }
 
     let libs = compilation.to_doc_test.iter().map(|package| {

